osm2streets/src/lib.rs

```diff
@@ -86,21 +87,36 @@ impl StreetNetwork {
         for i in [id.i1, id.i2] {
             self.intersections.get_mut(&i).unwrap().roads.push(id);
             self.sort_roads(i);
+            // Recalculate movements and complexity.
+            let (complexity, conflict_level, movements) =
+                crate::transform::classify_intersections::guess_complexity(self, &i);
```

Comment:
This sort of thing challenges my understanding of what should be a "transformation step" and what is an inherent part of the data types. Or at least which transformations are responsible for calculating data (that therefore need to be rerun when later transformations want to change things, such as merging intersections).

The "calculate change first, then commit them all" pattern that is forced upon us by the borrow checker in our current transformation implementations might actually be a better idea than calling this directly whenever a road is added. That is, maybe the responsibility for recalculating the movements etc. should be higher up than `insert_road`. At the level of `merge_intersections(a, b)` or higher.

Anyway, I've got a lot to understand yet about how the transformations fit together...

Reply:
> This sort of thing challenges my understanding of what should be a "transformation step" and what is an inherent part of the data types

I don't have clear ideas here either. :(  I'm going to try to draw the flow chart of how everything gets created / set currently. The big hammer is to treat this as a big dataflow-type problem and consider automatically recalculating derived state when something changes. Or not do that automatically, but at least have a principled way to organize / reason about this. The problem is that changes percolate to related objects. If you do anything to a road, you need to recalculate geometry for two intersections. That calculation might trim back other roads one hop away. If there are transformations that look at trimmed road length, would we want to always go trigger them again? It gets confusing very quickly...

> maybe the responsibility for recalculating the movements etc. should be higher up than insert_road. At the level of merge_intersections(a, b) or higher.

I like this way of thinking. Sometimes a very low-level, "surgical" set of mutations can happen. Or we use this "medium level" of abstraction. But another transformation may just want to say "merge short road" and automatically recalculate a bunch of derived state.

Reply:
> Or not do that automatically, but at least have a principled way to organize / reason about this [big dataflow-type problem].

This seems the likely best approach to me. At least for the code that takes the raw input to the naively "filled out" street representation.

My idea is that when you call `StreetNetwork::new(osm_stuff)` you might expect to get back `Road`s and `Intersection`s that are complete, but blindly representing the raw data. There would be a handful of low-level conditions that are met ("all the geometry connects up properly to each other..."), and the other higher level conditions haven't been addressed yet ("... but some of it overlaps"). These are the bits of code that don't feel like "transformations" to me, more like "hydrating" the abstract raw descriptions ("this centerline with this placement and these lanes") into a useful concrete representation ("these lanes covering these areas"). 

Because the raw data is criminally under-specified we are expected to infer a bunch of stuff; under-represented so we want to guess/fill in a bunch of stuff; and imprecise so we want to correct a bunch of stuff. For example, if all intersection-internal ways were tagged `junction=intersection` (or whatever the tag is) then we wouldn't need a transformation to merge intersections together, we'd simply collect all the connected intersection nodes together before creating the `Intersection`.

These inferences, guesses and corrections are easier to make with the *hydrated* representation, so it makes sense to *hydrate* the raw data, calculate any inferences, propagate them back into the raw data then *hydrate* again. This is how I see "transformations" fitting in conceptually.

Actually modifying the osm data as the result of transformations sounds like a painful idea, but it feels like there's something there: `StreetNetwork`/`Road`/`Intersection` are like primitives, with some source-of-truth data (placement adjusted centerlines, connectivity, stop line positions) and some straightforwardly derived data (geometry, etc.). If `StreetNetwork` provided a rich enough set of operations (like collapsing a road (merging intersections), collapsing an intersection (merging roads), zipping/snapping adjacent roads, adding turn restrictions, moving centerlines/stop line positions, that sort of thing) then it could be used to manually draw beautifully accurate street networks (or a horrible mess). The problem of getting a beautiful street network out of raw osm data then becomes one of naively translating the osm spec into `StreetNetwork` primitives, then doing bunch of tweaking operations on them using osm-specific inferences/assumptions.

The `StreetNetwork` primitive operations could be efficiently implemented, maybe even in a way that allows batching (perform a bunch of operations and keep track of what is invalidated, then recalculate all the invalidated data). Then we could reason about the order of transformations, if they need repeating, even of multiple transformations can be committed at the same time.

I'm gonna let that marinate in my brain for a little while and look to talk with you about it later.

Reply:
I just finished writing up #117 and this sounds so much better. The dataflow "framework" doesn't have to make the entire codebase complex; this batching executor could isolate it. I really agree at least two of the current transformations are something much more fundamental / less optional